File: algorithms/3_hard/M/00768_max_chunks_to_make_sorted_ii/stack.cpp
typedef ptrdiff_t isize;

class Solution {
public:
    static int maxChunksToSorted(vector<int>& arr) {
        const isize len = std::ssize(arr);
        isize stack_len = 0;
        int max = numeric_limits<int>::min();
        for (isize i = 0; i < len; i += 1) {
            max = std::max(max, arr[i]);
            while (stack_len > 0 && arr[stack_len - 1] > arr[i]) {
                stack_len -= 1;
            }
            arr[stack_len] = max;
            stack_len += 1;
        }
        assert(stack_len <= numeric_limits<int>::max());
        return static_cast<int>(stack_len);
    }
};

File: algorithms/3_hard/M/00768_max_chunks_to_make_sorted_ii/stack.rs
impl Solution {
    pub fn max_chunks_to_sorted(mut arr: Vec<i32>) -> i32 {
        let len = arr.len();
        let mut stack_len = 0_usize;
        let mut max = i32::MIN;
        for i in 0..len {
            max = max.max(arr[i]);
            while stack_len > 0 && arr[stack_len - 1] > arr[i] {
                stack_len -= 1;
            }
            arr[stack_len] = max;
            stack_len += 1;
        }
        stack_len.try_into().unwrap()
    }
}
